dart/dynamics/ShapeFrame.cpp

```diff
@@ -0,0 +1,394 @@
+bool VisualAddon::isHidden() const
+{
+  return getHidden();
+}
+
+//==============================================================================
+CollisionAddon::CollisionAddon(
+    common::AddonManager* mgr,
+    const AddonType::PropertiesData& properties)
+  : AddonType(mgr, properties)
+{
+//  mIsShapeNode = dynamic_cast<ShapeNode*>(mgr);
+}
+
+//==============================================================================
+bool CollisionAddon::isCollidable() const
+{
+  return getCollidable();
+}
+
+//==============================================================================
+ShapeFrame::UniqueProperties::UniqueProperties(const ShapePtr& shape)
+  : mShape(shape)
+{
+  // Do nothing
+}
+
+//==============================================================================
+ShapeFrame::UniqueProperties::UniqueProperties(ShapePtr&& shape)
+  : mShape(std::move(shape))
+{
+  // Do nothing
+}
+
+//==============================================================================
+ShapeFrame::Properties::Properties(
+    const Entity::Properties& entityProperties,
+    const UniqueProperties& shapeFrameProperties,
+    const ShapeFrame::AddonProperties& addonProperties)
+  : Entity::Properties(entityProperties),
+    UniqueProperties(shapeFrameProperties),
+    mAddonProperties(addonProperties)
+{
+  // Do nothing
+}
+
+//==============================================================================
+void ShapeFrame::setProperties(const ShapeFrame::AddonProperties& properties)
+{
+  setAddonProperties(properties);
+}
+
+//==============================================================================
+void ShapeFrame::setProperties(const Properties& properties)
+{
+  Entity::setProperties(static_cast<const Entity::Properties&>(properties));
+  setProperties(static_cast<const UniqueProperties&>(properties));
+  setProperties(properties.mAddonProperties);
+}
+
+//==============================================================================
+void ShapeFrame::setProperties(const ShapeFrame::UniqueProperties& properties)
+{
+  setShape(properties.mShape);
+}
+
+//==============================================================================
+const ShapeFrame::Properties ShapeFrame::getShapeFrameProperties() const
+{
+  return Properties(getEntityProperties(), mShapeFrameP, getAddonProperties());
+}
+
+//==============================================================================
+void ShapeFrame::copy(const ShapeFrame& other)
+{
+  if (this == &other)
+    return;
+
+  setProperties(other.getShapeFrameProperties());
+}
+
+//==============================================================================
+void ShapeFrame::copy(const ShapeFrame* other)
+{
+  if (nullptr == other)
+    return;
+
+  copy(*other);
+}
+
+//==============================================================================
+ShapeFrame& ShapeFrame::operator=(const ShapeFrame& other)
+{
+  copy(other);
+  return *this;
+}
```

Comment:
I'm not sure if it's a good idea to overload the assignment operator on this class. This is a pure virtual class, so all sorts of bad things could happen if you you try to assign two different `ShapeFrame` subclasses.